functions/src/extract.ts

```diff
@@ -20,49 +20,54 @@ import { DocumentSnapshot } from 'firebase-functions/lib/providers/firestore';
 import config from './config';
 import * as logs from './logs';
 import { dataProcessor, valueProcessor } from './processors';
+import transform from './transform';
 import { getObjectSizeInBytes, getFields, isValidValue } from './util';
 
 const PAYLOAD_MAX_SIZE = 102400;
 const PAYLOAD_TOO_LARGE_ERR_MSG = 'Record is too large.';
 const trim = /^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g;
 
-const getPayload = (snapshot: DocumentSnapshot) => {
-  const payload: {
+const getPayload = async (snapshot: DocumentSnapshot): Promise<any> => {
+  let payload: {
     [key: string]: boolean | string | number;
   } = {
-    objectID: snapshot.id,
+    // adding the objectId for use in Post processing.  The objectId is only meant to be read-only.
+    //  any changes will to objectId will not be used and will be restored in the return statement.
+    objectID: snapshot.id
   };
 
   const fields = getFields(config);
   if (fields.length === 0) {
-    return {
+    payload = {
       ...dataProcessor(snapshot.data()),
-      ...payload,
+      ...payload
     };
-  }
-
-  // Fields have been defined by user.  Start pulling data from the document to create payload
-  // to send to Algolia.
-  fields.forEach(item => {
-    let firebaseField = item.replace(trim, '');
-    const [field, value] = valueProcessor(firebaseField, snapshot.get(firebaseField));
+  } else {
 
-    logs.debug('field', field);
-    logs.debug('value', value);
+    // Fields have been defined by user.  Start pulling data from the document to create payload
+    // to send to Algolia.
+    fields.forEach(item => {
+      let firebaseField = item.replace(trim, '');
+      const [ field, value ] = valueProcessor(firebaseField, snapshot.get(firebaseField));
 
-    if (isValidValue(value)) {
-      payload[field] = value;
-    } else {
-      logs.fieldNotExist(firebaseField);
-    }
-  });
+      if (isValidValue(value)) {
+        payload[field] = value;
+      } else {
+        logs.fieldNotExist(firebaseField);
+      }
+    });
+  }
 
-  return payload;
+  // adding the objectId in the return to make sure to restore to original if changed in the post processing.
+  return {
+    ...await transform(payload),
+    objectID: snapshot.id,
```

Comment:
what if you wanted to transform the objectID, in case of splitting a long record for example?

Reply:
Should transform allow an array of objects as output?